README.md
lazBiorhythms

A Lazarus program to draw Biorhythms.

The Biorhythms for one or two people can be drawn, just click on the appropriate user.

Once a Birth date has been entered, the total number of days alive will be displayed.

The following rhythms can be displayed - 

Primary
	Physical, with a period of 23 days
	Emotional, with a period of 28 days
	Intellectual, with a period of 33 days

Secondary
	Spiritual, with a period of 53 days
	Awareness, with a period of 48 days
	Aesthetic, with a period of 43 days
	Intuition, with a period of 38 days

Both the Primary and Secondary can be combined into one plot, as can all seven rhythms.

The colours of all the plots can be customised, go to file / options.

Once Birth dates have been entered and change of colours, will be remembered and used next time.



history.txt
Build 1

Initial import into git.
Basic layout of form complete.
Basic programming code complete.

The program accepts a birthday and draws the Biorhythms.

Needs adding to and making more robust.


Build 2
Changed the checkGroups for individual check boxes, mainly so that the text of each could match the colour of the line series.

Added Primary and Secondary Biorhythms

Currently
262 lines of source code.
240 statements
14.1% 


Build 3

Added pre-built forms for Help, Licence, About and Options.
Added pre-built units for utils and options.
Added menus to main form to call the above.

Currently
1348 lines of source code.
649 statements
18.5% 


Build 4

Added user options for a second user.
Added a second TDateEdit to enter birthrate of second user.

Currently
1752 lines of source code.
956 statements
16.3% 


Build 5

Added the drawing stuff for the second user.

Moved plotChart into a separate procedure, this either draws each user's
curves or clear series if neither user is selected.

Added a new procedure setToday, to draw a vertical bar to indicate today, on the chart.
It looks at each user series to try and determine the range.
The secondary combined (if used) has the greatest range.
If both user's combines are displayed, use the min and max of the two.
If neither combined is used, use defaults..

Currently
1897 lines of source code.
1061 statements
16.1% 


Build 6

Added a plot for total of all seven rhythms [overall].

Set up colour stuff for second user.

Only enable stuff for a user, if that user is selected.

Corrected range for the today marker [see procedure setToday].

Currently
1983 lines of source code.
1139 statements
15.6% 


Build 7

Completed the code and testing of the colour option for both people.

Combined all the colour button colour change procedures into one generic
procedure in formOptions.

Corrected a bug with writing default options in uOptions.

Corrected a bug with compile time in formAbout.

Currently
1982 lines of source code.
1180 statements
16.5% 


Build 8

Corrected memory leaks in uOptions.pas - not freeing fvi object.

Added constants to frmLazBiorhythms.pas for periods of the rhythms.

Corrected spelling.

** Compiled with Lazarus V2.0.0.4 **

Currently
2015 lines of source code.
1192 statements
16.2% 


formhelp.pas
unit formhelp;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, Forms, Controls, Graphics, Dialogs, StdCtrls,
  ExtCtrls;

type

  { TfrmHelp }

  TfrmHelp = class(TForm)
    btnhelpExit : TButton;
    lblComments : TLabel;
    lblCopyRight: TLabel;
    lblVersion  : TLabel;
    mmoHelp     : TMemo;
    Panel1      : TPanel;
    Panel2      : TPanel;

    procedure btnhelpExitClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { private declarations }
  public
    { public declarations }
  end; 

var
  frmHelp : TfrmHelp;

implementation

uses
  formLazBiorhythms;

{$R *.lfm}

{ TfrmHelp }

procedure TfrmHelp.FormCreate(Sender: TObject);
begin
  mmoHelp.Append('LazBiorhythms.');
  mmoHelp.Append('');
  mmoHelp.Append('');

  try
    mmoHelp.Lines.LoadFromFile('help.txt');
  except
    on Exception do begin
      mmoHelp.Append(' help file not found.');
      mmoHelp.Append('');
      mmoHelp.Append(' This file should include full and detailed help instructions.');
    end;
  end;

  lblComments.Caption  := format('%s', [userOptions.Comments]);
  lblCopyRight.Caption := userOptions.legalCopyright;
  lblVersion.Caption   := format('%s Version :: %s', [userOptions.productName, userOptions.fileVersion]);
end;

procedure TfrmHelp.btnhelpExitClick(Sender: TObject);
begin
  Close;
end;


end.



formabout.pas
unit formAbout;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, Forms, Controls, Graphics, Dialogs, StdCtrls,
  LCLVersion, ExtCtrls, ShellApi, LazBiorhythmsUntils;

type

  { TfrmAbout }

  TfrmAbout = class(TForm)
    btnAboutExit         : TButton;
    btnAboutMSInfo       : TButton;
    Image1               : TImage;
    lblProgramName       : TLabel;
    lblProgramDescription: TLabel;
    lblProgrammer        : TLabel;
    lblSysUpTime         : TLabel;
    lblAppUpTime         : TLabel;
    lblSystemUpTime      : TLabel;
    lblApplicationUpTime : TLabel;
    LstBxInfo            : TListBox;
    LstBxDiscSpace       : TListBox;
    Panel1               : TPanel;
    Panel2               : TPanel;
    Panel3               : TPanel;
    Panel4               : TPanel;
    tmrUpTime            : TTimer;

    procedure btnAboutExitClick(Sender: TObject);
    procedure btnAboutMSInfoClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure tmrUpTimeTimer(Sender: TObject);
  private
    { private declarations }
  public
    { public declarations }
  end; 

var
  frmAbout: TfrmAbout;

implementation

uses
  formLazBiorhythms;

{$R *.lfm}

{ TfrmAbout }


procedure TfrmAbout.btnAboutExitClick(Sender: TObject);
{  Close About form.  }
begin
  tmrUpTime.Enabled := false;
  Close;
end;

procedure TfrmAbout.btnAboutMSinfoClick(Sender: TObject);
{  Run the external application MSinfo.
   If ShellExecute returns 32 or less this indicates an error, just inform the user.
}
begin
  if ShellExecute(0, nil, PChar('"msinfo32.exe"'), nil, nil, 1) < 33 then
    ShowMessage('ERROR : running MSinfo');
end;

procedure TfrmAbout.FormCreate(Sender: TObject);
var
  dskSize  : string;
  dskFree  : string;
  message  : string;
  cmpDate  : string;
  cmpUkDate: string;
  i        : integer;

begin
  tmrUpTime.Enabled     := True;
  lblAppUpTime.Caption  := getUpTime('Application');
  lblSysUpTime.Caption  := getUpTime('System');
  lblProgrammer.Caption := userOptions.legalCopyright;

  //  {$I %DATE%} returns the compile date, but in American [ignores local date format]
  //  So, we string slice it to give good old English date format.
  cmpDate               := {$I %DATE%};
  cmpUkDate             := format('%s/%s/%s', [copy(cmpDate, 9, 2),
                                               copy(cmpDate, 6, 2),
                                               copy(cmpDate, 1, 4)]);

  lstBxInfo.Items.add(userOptions.fileDescription);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(format('lazBiorhythms Version :: %s', [userOptions.fileVersion]));
  lstBxInfo.Items.add(format('lazBiorhythms Built   :: %s', [cmpUkDate + ' @ ' + {$I %TIME%}]));

  {$ifdef WIN32}
    lstBxInfo.Items.add(format('Built with 32 bit Lazarus Version :: %s', [lcl_version]));
  {$else}
    lstBxInfo.Items.add(format('Built with 64 bit Lazarus Version :: %s', [lcl_version]));
  {$endif}
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(getWindowsVersion);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(userOptions.CompanyName);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add(userOptions.Comments);
  lstBxInfo.Items.add('');
  lstBxInfo.Items.add('App Dir : ' + ExtractFilePath(Application.ExeName));

  // Display the free space on drives B, C, D, E, F, where present
  for i := 2 to 10 do
  begin
    dskFree := FloatToStrF(DiskFree(i) / 1073741824, ffFixed, 3, 2);
    dskSize := FloatToStrF(DiskSize(i) / 1073741824, ffFixed, 3, 2);

    if DiskSize(i) >= 0 then
    begin
      message := format(' Disk %s : Free / Size :: %s / %s Gbytes', [Chr(i + 64), dskFree, dskSize]);
      LstBxDiscSpace.Items.Add(message);
    end;
  end;

end;

procedure TfrmAbout.tmrUpTimeTimer(Sender: TObject);
{  Update the labels in real time.    }
begin
  lblAppUpTime.Caption := getUpTime('Application');
  lblSysUpTime.Caption := getUpTime('System');
end;

end.



formlicence.pas
unit formLicence;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, Forms, Controls, Graphics, Dialogs, StdCtrls,
  ExtCtrls;

type

  { TfrmLicence }

  TfrmLicence = class(TForm)
    btnLicenceExit: TButton;
    lblComments   : TLabel;
    lblCopyRight  : TLabel;
    lblVersion    : TLabel;
    mmoLicence    : TMemo;
    Panel1        : TPanel;
    Panel2        : TPanel;

    procedure btnLicenceExitClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
  private
    { private declarations }
  public
    { public declarations }
  end; 

var
  frmLicence: TfrmLicence;

implementation

uses
  formLazBiorhythms;

{$R *.lfm}

{ TfrmLicence }

procedure TfrmLicence.FormCreate(Sender: TObject);
begin
  mmoLicence.Append('LazBiorhythms.');
  mmoLicence.Append('');
  try
    mmoLicence.Lines.LoadFromFile('GNU GENERAL PUBLIC LICENSE.txt');
  except
    on Exception do begin
      mmoLicence.Append(' help License not found.');
      mmoLicence.Append('');
      mmoLicence.Append(' The application is issued under the GNU GENERAL PUBLIC LICENSE.');
    end;
  end;

  lblComments.Caption  := userOptions.Comments;
  lblCopyRight.Caption := userOptions.legalCopyright;
  lblVersion.Caption   := format('%s Version :: %s', [userOptions.productName, userOptions.fileVersion]);
end;

procedure TfrmLicence.btnLicenceExitClick(Sender: TObject);
begin
  Close;
end;

end.



formlazbiorhythms.pas
unit formLazBiorhythms;

{
    A Lazarus program to draw a persons Biorhythms
    Copyright (C) <2018>  <Kevin Scott>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
}

{
Cycle           Length
Physical 	23 days *
Emotional 	28 days *
Intellectual	33 days *
Spiritual       53 days
Awareness	48 days
Aesthetic	43 days
Intuition	38 days

* Primary curves
}


{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, FileUtil, TAGraph, TAIntervalSources, TASeries, Forms,
  Controls, Graphics, Dialogs, StdCtrls, ExtCtrls, EditBtn, Menus, dateutils,
  formhelp, formLicence, formAbout, formOptions, uOptions, math;

CONST
  PHYSICAL     = 23;
  EMOTIONAL    = 28;
  INTELLECTUAL = 33;
  SPIRITUAL    = 53;
  AWARENESS    = 48;
  AESTHETIC    = 43;
  INTUITION    = 38;
  PLOT_START   = 30;     //  where to start the plot.

type

  { TfrmLazBiorhythms }

  TfrmLazBiorhythms = class(TForm)
    chckBxFirstAesthetic         : TCheckBox;
    chckBxFirstOverall           : TCheckBox;
    chckBxSecondAesthetic        : TCheckBox;
    chckBxFirstAwareness         : TCheckBox;
    chckBxSecondAwareness        : TCheckBox;
    chckBxSecondEmotional        : TCheckBox;
    chckBxSecondIntellectual     : TCheckBox;
    chckBxFirstIntuition         : TCheckBox;
    chckBxSecondIntuition        : TCheckBox;
    chckBxSecondPhysical         : TCheckBox;
    chckBxSecondPrimaryCombined  : TCheckBox;
    chckBxFirstSecondaryCombined : TCheckBox;
    chckBxSecondSecondaryCombined: TCheckBox;
    chckBxFirstSpirtual          : TCheckBox;
    chckBxSecondSpirtual         : TCheckBox;
    chckBxFirstPhysical          : TCheckBox;
    chckBxFirstIntellectual      : TCheckBox;
    chckBxFirstEmotional         : TCheckBox;
    chckBxFirstPrimaryCombined   : TCheckBox;
    chckBxSecondUser             : TCheckBox;
    chckBxFirstUser              : TCheckBox;
    chckBxSecondOverall          : TCheckBox;
    ChrtBiorhythms               : TChart;
    DateTimeIntervalChartSource1 : TDateTimeIntervalChartSource;
    lnSrsFirstAwareness          : TLineSeries;
    lnSrsFirstAesthetic          : TLineSeries;
    lnSrsFirstSecondaryCombined  : TLineSeries;
    lnSrsFirstIntuition          : TLineSeries;
    lnSrsFirstSpiritial          : TLineSeries;
    lnSrsFirstPrimaryCombined    : TLineSeries;
    lnSrsFirstIntellectual       : TLineSeries;
    lnSrsFirstEmotional          : TLineSeries;
    lnSrsFirstPhysical           : TLineSeries;
    lnSrsSecondSecondaryCombined : TLineSeries;
    lnSrsSecondAesthetic         : TLineSeries;
    lnSrsSecondIntuition         : TLineSeries;
    lnSrsSecondAwareness         : TLineSeries;
    lnSrsSecondSpiritial         : TLineSeries;
    lnSrsSecondPrimaryCombined   : TLineSeries;
    lnSrsSecondIntellectual      : TLineSeries;
    lnSrsSecondEmotional         : TLineSeries;
    lnSrsSecondPhysical          : TLineSeries;
    lnSrsFirstOverall            : TLineSeries;
    lnSrsSecondOverall           : TLineSeries;
    lnSrsTodayMark               : TBarSeries;
    lblFirstAesthetic            : TLabel;
    lblSecondAesthetic           : TLabel;
    lblFirstAwareness            : TLabel;
    lblSecondAwareness           : TLabel;
    lblSecondCombined            : TLabel;
    lblSecondEmotional           : TLabel;
    lblSecondIntellectual        : TLabel;
    lblFirstIntuition            : TLabel;
    lblSecondIntuition           : TLabel;
    lblSecondPhysical            : TLabel;
    lblFirstSecondaryCombined    : TLabel;
    lblSecondSecondaryCombined   : TLabel;
    lblFirstSpiritual            : TLabel;
    lblSecondSpiritual           : TLabel;
    lblFirstOverall              : TLabel;
    lblSecondOverall             : TLabel;
    lblFirstCombined             : TLabel;
    lblFirstEmotional            : TLabel;
    lblFirstIntellectual         : TLabel;
    lblFirstPhysical             : TLabel;
    lblFirstBirthdayInfo         : TLabel;
    lblSecondBirthdayInfo        : TLabel;
    DtEdtFirstBirthDay           : TDateEdit;
    DtEdtSecondBirthDay          : TDateEdit;
    MnMenuBio                    : TMainMenu;
    MenuItmOptions               : TMenuItem;
    MenuItmExit                  : TMenuItem;
    MnIimFile                    : TMenuItem;
    MenuItmHelp                  : TMenuItem;
    MenuItmLicence               : TMenuItem;
    MenuItmAbout                 : TMenuItem;
    MnItmHelp                    : TMenuItem;
    Panel1                       : TPanel;
    btnExit                      : TButton;
    GrpBxBirthDate               : TGroupBox;
    GrpBxFirstUser               : TGroupBox;
    GrpBxSecondUser              : TGroupBox;

    procedure btnExitClick(Sender: TObject);
    procedure chckBxFirstUserChange(Sender: TObject);
    procedure chckBxSecondUserChange(Sender: TObject);
    procedure DtEdtFirstBirthDayChange(Sender: TObject);
    procedure DtEdtSecondBirthDayChange(Sender: TObject);
    procedure FormClose(Sender: TObject; var CloseAction: TCloseAction);
    procedure FormCreate(Sender: TObject);
    procedure checkBoxClick(Sender: TObject);
    procedure MenuItmAboutClick(Sender: TObject);
    procedure MenuItmExitClick(Sender: TObject);
    procedure MenuItmHelpClick(Sender: TObject);
    procedure MenuItmLicenceClick(Sender: TObject);
    procedure MenuItmOptionsClick(Sender: TObject);
  private
    procedure plotChart;
    procedure clearSeries;
    procedure plotSeries;
    procedure colourStuff;
    procedure setToday;

  public

  end;

var
  frmLazBiorhythms: TfrmLazBiorhythms;
  userOptions     : Options;              //  used to hold user options.
  appStartTime    : int64;                //  used by formAbout to determine how long the app has been running.

implementation

{$R *.lfm}

{ TfrmLazBiorhythms }

procedure TfrmLazBiorhythms.FormCreate(Sender: TObject);
{  Set up stuff at form create.    }
begin
  appStartTime := GetTickCount64;  //  tick count when application starts.

  userOptions  := Options.Create;  // create options file as c:\Users\<user>\AppData\Local\LazBiorhythms\Options.xml

  lblFirstBirthdayInfo.Caption  := 'Please enter your birthday';
  lblSecondBirthdayInfo.Caption := 'Please enter your birthday';

  DtEdtFirstBirthDay.Date        := userOptions.Firstbirthdate;
  DtEdtSecondBirthDay.Date       := userOptions.Secondbirthdate;
  DtEdtFirstBirthDay.Font.Color  := clblack;
  DtEdtSecondBirthDay.Font.Color := clblack;

  chckBxFirstUser.Checked  := userOptions.useFirstUser;
  chckBxSecondUser.Checked := userOptions.useSecondUser;
end;

procedure TfrmLazBiorhythms.FormClose(Sender: TObject; var CloseAction: TCloseAction);
begin
  userOptions.writeCurrentOptions;  // write out options file.
  userOptions.Free;                 //  Release the user options.
  CloseAction:= caFree;
end;

procedure TfrmLazBiorhythms.DtEdtFirstBirthDayChange(Sender: TObject);
{  When a first birth date has been entered, draw chart.    }
begin
  userOptions.Firstbirthdate   := DtEdtFirstBirthDay.Date;
  lblFirstBirthdayInfo.Caption := format('You have been alive %d days', [DaysBetween(today, userOptions.Firstbirthdate)]);

  plotSeries
end;

procedure TfrmLazBiorhythms.DtEdtSecondBirthDayChange(Sender: TObject);
{  When a second birth date has been entered, draw chart.    }
begin
  userOptions.Secondbirthdate   := DtEdtSecondBirthDay.Date;
  lblSecondBirthdayInfo.Caption := format('You have been alive %d days', [DaysBetween(today, userOptions.Secondbirthdate)]);

  plotSeries
end;

procedure TfrmLazBiorhythms.btnExitClick(Sender: TObject);
{  Close the app.    }
begin
  close;
end;

procedure TfrmLazBiorhythms.plotChart;
{  Draw the actual Biorhythms.

   First and second user things need a tidy up.
}
const
  N = 60;                             //  number of days.
var
  f         : Integer;
  x         : Double;
  sdays     : integer;
  days      : double;
  sdate     : TdateTime;
  daysAlive : integer;
begin
  clearSeries;
  colourStuff;

  if userOptions.useFirstUser  then
  begin
    daysAlive := DaysBetween(today, userOptions.Firstbirthdate);

    sdays := daysAlive - PLOT_START;        //  where to start the plot.
    sdate := today;
    sdate := sdate - PLOT_START;

    for f:=0 to N-1 do begin
      x    := sdays + f;
      days := 2 * PI * x;

      if chckBxFirstPhysical.Checked then
        lnSrsFirstPhysical.AddXY(sdate, sin(days / PHYSICAL));
      if chckBxFirstEmotional.Checked
        then lnSrsFirstEmotional.AddXY(sdate, sin(days / EMOTIONAL));
      if chckBxFirstIntellectual.Checked
        then lnSrsFirstIntellectual.AddXY(sdate, sin(days / INTELLECTUAL));
      if chckBxFirstPrimaryCombined.Checked then
        lnSrsFirstPrimaryCombined.AddXY(sdate, sin(days / PHYSICAL) +
                                               sin(days / EMOTIONAL) +
                                               sin(days / INTELLECTUAL));

      if chckBxFirstSpirtual.Checked then
        lnSrsFirstSpiritial.AddXY(sdate, sin(days / SPIRITUAL));
      if chckBxFirstAwareness.Checked then
        lnSrsFirstAwareness.AddXY(sdate, sin(days / AWARENESS));
      if chckBxFirstAesthetic.Checked then
        lnSrsFirstAesthetic.AddXY(sdate, sin(days / AESTHETIC));
      if chckBxFirstIntuition.Checked then
        lnSrsFirstIntuition.AddXY(sdate, sin(days / INTUITION));
      if chckBxFirstSecondaryCombined.Checked then
        lnSrsFirstSecondaryCombined.AddXY(sdate, sin(days / SPIRITUAL) +
                                                 sin(days / AWARENESS) +
                                                 sin(days / AESTHETIC) +
                                                 sin(days / INTUITION));

      if chckBxFirstOverall.Checked then
        lnSrsFirstOverall.AddXY(sdate, sin(days / PHYSICAL) +
                                       sin(days / EMOTIONAL) +
                                       sin(days / INTELLECTUAL) +
                                       sin(days / SPIRITUAL) +
                                       sin(days / AWARENESS) +
                                       sin(days / AESTHETIC) +
                                       sin(days / INTUITION));
      sdate := sdate + 1;
    end;
  end;

  if userOptions.useSecondUser  then
  begin
    daysAlive := DaysBetween(today, userOptions.Secondbirthdate);

    sdays := daysAlive - PLOT_START;        //  where to start the plot.
    sdate := today;
    sdate := sdate - PLOT_START;

    for f:=0 to N-1 do begin
      x    := sdays + f;
      days := 2 * PI * x;

      if chckBxSecondPhysical.Checked then
        lnSrsSecondPhysical.AddXY(sdate, sin(days / PHYSICAL));
      if chckBxSecondEmotional.Checked
        then lnSrsSecondEmotional.AddXY(sdate, sin(days / EMOTIONAL));
      if chckBxSecondIntellectual.Checked
        then lnSrsSecondIntellectual.AddXY(sdate, sin(days / INTELLECTUAL));
      if chckBxSecondPrimaryCombined.Checked then
        lnSrsSecondPrimaryCombined.AddXY(sdate, sin(days / PHYSICAL) +
                                                sin(days / EMOTIONAL) +
                                                sin(days / INTELLECTUAL));

      if chckBxSecondSpirtual.Checked then
        lnSrsSecondSpiritial.AddXY(sdate, sin(days / SPIRITUAL));
      if chckBxSecondAwareness.Checked then
        lnSrsSecondAwareness.AddXY(sdate, sin(days / AWARENESS));
      if chckBxSecondAesthetic.Checked then
        lnSrsSecondAesthetic.AddXY(sdate, sin(days / AESTHETIC));
      if chckBxSecondIntuition.Checked then
        lnSrsSecondIntuition.AddXY(sdate, sin(days / INTUITION));
      if chckBxSecondSecondaryCombined.Checked then
        lnSrsSecondSecondaryCombined.AddXY(sdate, sin(days / SPIRITUAL) +
                                                  sin(days / AWARENESS) +
                                                  sin(days / AESTHETIC) +
                                                  sin(days / INTUITION));

      if chckBxSecondOverall.Checked then
        lnSrsSecondOverall.AddXY(sdate, sin(days / PHYSICAL) +
                                        sin(days / EMOTIONAL) +
                                        sin(days / INTELLECTUAL) +
                                        sin(days / SPIRITUAL) +
                                        sin(days / AWARENESS) +
                                        sin(days / AESTHETIC) +
                                        sin(days / INTUITION));

      sdate := sdate + 1;
    end;
  end;

  setToday;

  DtEdtFirstBirthDay.Date  := userOptions.Firstbirthdate;
  DtEdtSecondBirthDay.Date := userOptions.Secondbirthdate;
end;

procedure TfrmLazBiorhythms.setToday;
{  Draws a vertical bar to indicate today, on the chart.

   It looks at each user series to try and determine the range.
   The Overall series will be the larger, then the combined.
   The individual series all will have the same magnitude.
   If either Overall or Combined are not used - they return inf, which if ignored.
   All returned values are loaded into a array, then min or max can be determined.
   If none is used, use defaults.
}
Type
  values = array[0..4] of double;
VAR
  minArray : values;
  maxArray : values;
begin
  minArray[0] := lnSrsFirstSecondaryCombined.MinYValue;
  minArray[1] := lnSrsSecondSecondaryCombined.MinYValue;
  minArray[2] := lnSrsFirstOverall.MinYValue;
  minArray[3] := lnSrsSecondOverall.MinYValue;
  minArray[4] := -2.5;                                  //  use defaults.

  maxArray[0] := lnSrsFirstSecondaryCombined.MaxYValue;
  maxArray[1] := lnSrsSecondSecondaryCombined.MaxYValue;
  maxArray[2] := lnSrsFirstOverall.MaxYValue;
  maxArray[3] := lnSrsSecondOverall.MaxYValue;
  maxArray[4] := 2.5;                                   //  use defaults

  //  draw the verticsal today line.
  lnSrsTodayMark.ZeroLevel := minvalue(minArray);
  lnSrsTodayMark.AddXY(today, maxvalue(maxArray));
end;

procedure TfrmLazBiorhythms.clearSeries;
{  Clears the chart's line series.    }
begin
  lnSrsFirstPhysical.Clear;
  lnSrsFirstEmotional.Clear;
  lnSrsFirstIntellectual.Clear;
  lnSrsFirstPrimaryCombined.Clear;

  lnSrsFirstSpiritial.Clear;
  lnSrsFirstAwareness.Clear;
  lnSrsFirstAesthetic.Clear;
  lnSrsFirstIntuition.Clear;
  lnSrsFirstSecondaryCombined.Clear;

  lnSrsFirstOverall.Clear;


  lnSrsSecondPhysical.Clear;
  lnSrsSecondEmotional.Clear;
  lnSrsSecondIntellectual.Clear;
  lnSrsSecondPrimaryCombined.Clear;

  lnSrsSecondSpiritial.Clear;
  lnSrsSecondAwareness.Clear;
  lnSrsSecondAesthetic.Clear;
  lnSrsSecondIntuition.Clear;
  lnSrsSecondSecondaryCombined.Clear;

  lnSrsSecondOverall.Clear;


  lnSrsTodayMark.Clear;
end;

procedure TfrmLazBiorhythms.plotSeries;
{  If either user is checked then draw their curves,
   if neither then clear series.
}
begin
  if (userOptions.useFirstUser or userOptions.useSecondUser) then
    plotChart
  else
    clearSeries;
end;

procedure TfrmLazBiorhythms.colourStuff;
{  Sets up default colours for the labels on line series.    }
begin

  lnSrsFirstPhysical.SeriesColor          := userOptions.clrFirstPhysical;
  lblFirstPhysical.Font.Color             := userOptions.clrFirstPhysical;
  lnSrsFirstIntellectual.SeriesColor      := userOptions.clrFirstIntellectual;
  lblFirstIntellectual.Font.Color         := userOptions.clrFirstIntellectual;
  lnSrsFirstEmotional.SeriesColor         := userOptions.clrFirstEmotional;
  lblFirstEmotional.Font.Color            := userOptions.clrFirstEmotional;
  lnSrsFirstPrimaryCombined.SeriesColor   := userOptions.clrFirstPriCombined;
  lblFirstCombined.Font.Color             := userOptions.clrFirstPriCombined;
  lnSrsFirstSpiritial.SeriesColor         := userOptions.clrFirstSpitual;
  lblFirstSpiritual.Font.Color            := userOptions.clrFirstSpitual;
  lnSrsFirstAwareness.SeriesColor         := userOptions.clrFirstAwareness;
  lblFirstAwareness.Font.Color            := userOptions.clrFirstAwareness;
  lnSrsFirstAesthetic.SeriesColor         := userOptions.clrFirstAesthetic;
  lblFirstAesthetic.Font.Color            := userOptions.clrFirstAesthetic;
  lnSrsFirstIntuition.SeriesColor         := userOptions.clrFirstIntuition;
  lblFirstIntuition.Font.Color            := userOptions.clrFirstIntuition;
  lnSrsFirstSecondaryCombined.SeriesColor := userOptions.clrFirstSecCombined;
  lblFirstSecondaryCombined.Font.Color    := userOptions.clrFirstSecCombined;
  lnSrsFirstOverall.SeriesColor           := userOptions.clrFirstOverall;
  lblFirstOverall.Font.Color              := userOptions.clrFirstOverall;

  lblSecondPhysical.Font.Color             := userOptions.clrSecondPhysical;
  lnSrsSecondPhysical.SeriesColor          := userOptions.clrSecondPhysical;
  lblSecondIntellectual.Font.Color         := userOptions.clrSecondIntellectual;
  lnSrsSecondIntellectual.SeriesColor      := userOptions.clrSecondIntellectual;
  lblSecondEmotional.Font.Color            := userOptions.clrSecondEmotional;
  lnSrsSecondEmotional.SeriesColor         := userOptions.clrSecondEmotional;
  lblSecondCombined.Font.Color             := userOptions.clrSecondPriCombined;
  lnSrsSecondPrimaryCombined.SeriesColor   := userOptions.clrSecondPriCombined;
  lblSecondSpiritual.Font.Color            := userOptions.clrSecondSpitual;
  lnSrsSecondSpiritial.SeriesColor         := userOptions.clrSecondSpitual;
  lblSecondAwareness.Font.Color            := userOptions.clrSecondAwareness;
  lnSrsSecondAwareness.SeriesColor         := userOptions.clrSecondAwareness;
  lblSecondAesthetic.Font.Color            := userOptions.clrSecondAesthetic;
  lnSrsSecondAesthetic.SeriesColor         := userOptions.clrSecondAesthetic;
  lblSecondIntuition.Font.Color            := userOptions.clrSecondIntuition;
  lnSrsSecondIntuition.SeriesColor         := userOptions.clrSecondIntuition;
  lblSecondSecondaryCombined.Font.Color    := userOptions.clrSecondSecCombined;
  lnSrsSecondSecondaryCombined.SeriesColor := userOptions.clrSecondSecCombined;
  lnSrsSecondOverall.SeriesColor           := userOptions.clrSecondOverall;
  lblSecondOverall.Font.Color              := userOptions.clrSecondOverall;
end;
//
//........................................ Check Boxes .........................
//
procedure TfrmLazBiorhythms.checkBoxClick(Sender: TObject);
{  A generic event handler called whenever any of the checkboxes are clicked.
   Just calls the plot routine i.e. re-draws if a series is added or removed.
}
begin
  plotSeries
end;

procedure TfrmLazBiorhythms.chckBxFirstUserChange(Sender: TObject);
begin
  userOptions.useFirstUser     := chckBxFirstUser.Checked;
  GrpBxFirstUser.Enabled       := chckBxFirstUser.Checked;
  lblFirstBirthdayInfo.Enabled := chckBxFirstUser.Checked;
  DtEdtFirstBirthDay.Enabled   := chckBxFirstUser.Checked;

  plotSeries
end;

procedure TfrmLazBiorhythms.chckBxSecondUserChange(Sender: TObject);
begin
  userOptions.useSecondUser     := chckBxSecondUser.Checked;
  GrpBxSecondUser.Enabled       := chckBxSecondUser.Checked;
  lblSecondBirthdayInfo.Enabled := chckBxSecondUser.Checked;
  DtEdtSecondBirthDay.Enabled   := chckBxSecondUser.Checked;
  plotSeries
end;
//
//........................................ File Menu ...........................
//
procedure TfrmLazBiorhythms.MenuItmOptionsClick(Sender: TObject);
begin
  frmOptions.ShowModal;

  //  in case something has changed.
  colourStuff;
  plotSeries
end;

procedure TfrmLazBiorhythms.MenuItmExitClick(Sender: TObject);
begin
  userOptions.writeCurrentOptions;
  close;
end;
//
//........................................ Help Menu ...........................
//
procedure TfrmLazBiorhythms.MenuItmHelpClick(Sender: TObject);
begin
  frmHelp.ShowModal;
end;

procedure TfrmLazBiorhythms.MenuItmLicenceClick(Sender: TObject);
begin
  frmLicence.ShowModal;
end;

procedure TfrmLazBiorhythms.MenuItmAboutClick(Sender: TObject);
begin
  frmAbout.ShowModal;
end;
end.

